### PR TITLE
Set up linting / formatting / type checking / tests / CI

### LINKED ISSUE #23: Proposal: add linting / type checking / formatting / CI
Currently, this project has type hints but these are not checked. The lack of checking makes it easy for the types to accidentally be incorrect, and for bugs to slip through.

This project also has both `tool.ruff` and `tool.black` lines in the config, but it seems like neither formatter is installed. IMO ruff would be the better tool as it can also lint as well as formatting.

Finally, all this should be run in CI via Github Actions, so CI can automatically assert that tests pass, types are valid, and everything passes formatting / linting checks.

If this is acceptable, I can open a PR with these implemented.

### PR DESCRIPTION
**Pull request by chanind**
This PR sets up linting / formatting with ruff, testing with pytest, and type checking with pyright. This PR also runs all of these via Github Actions on PR and push to main.

Some notes:
- I used the loosest setting for Pyright (basic), since #23 indicated a dislike for strict type checking. The strictness can be increased in the future if desired.
- I used `# type: ignore` comments liberally. There were a number of places I did this that looked like legitimate bugs in the code, but I didn't want to make any functional changes if there was any chance I was wrong, especially until there are tests.
- When the declared types were just wrong (e.g., JumpReLU claiming that the threshold is a `float`), I changed the types so they are correct.
- I didn't use a dependency manager, so it's up to the user to figure out how to install dev deps correctly and make sure they have a venv and whatever

Closes #23

### CODE REVIEW

File: tests/test_attributions_gemma.py [hunk 1]
```diff
@@ -374,16 +376,19 @@ def verify_gemma_2_2b(s: str):
         verify_feature_edges(model, graph)
 
 
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
```
Comment (mntss) [inline tests/test_attributions_gemma.py:379]: I think we should be able to run the small / large tests on the CPU as a part of CI. These are the main tests for circuit tracer so including them in the CI would be pretty valuable. We definitely need to skip the full gemma test though

File: tests/test_graph.py [hunk 2]
```diff
@@ -1,11 +1,13 @@
 import numpy as np
+import pytest
 import torch
 from torch import device
 from transformer_lens import HookedTransformerConfig
 
 from circuit_tracer.graph import Graph, compute_edge_influence, compute_node_influence
 
 
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
```
Comment (mntss) [inline tests/test_graph.py:10]: This should be runnable without cuda

File: tests/test_attributions_llama.py [hunk 3]
```diff
@@ -200,23 +202,19 @@ def verify_llama_3_2_1b(s: str):
...
     verify_large_llama_model(s)
 
 
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
 def test_llama_3_2_1b():
     s = "The National Digital Analytics Group (ND"
     verify_llama_3_2_1b(s)
-
-
-if __name__ == "__main__":
```
Comment (mntss) [inline tests/test_attributions_llama.py:218]: I'm in favor of keeping these entrypoints, less friction for debugging

File: README.md [hunk 4]
```diff
@@ -116,15 +135,45 @@ circuit-tracer attribute \
...
+
 When using the `--server` option, your browser will open to a local visualization interface. The interface is the same as in [the original papers](https://transformer-circuits.pub/2025/attribution-graphs/methods.html) (frontend available [here](https://github.com/anthropics/attribution-graphs-frontend)).
+
 - **Select a node**: Click on a node.
 - **Pin / unpin a node to subgraph pane**: Ctrl+click/Commmand+click the node.
 - **Annotate a node**: Click on the "Edit" button on the right side of the window while a node is selected to edit its annotation.
 - **Group nodes**: Hold G and click on nodes to group them together into a supernode. Hold G and click on the x next to a supernode to ungroup all of them.
 - **Annotate supernode / node group**: click on the label below the supernode to edit the supernode annotation.
 
+### Development
```
Comment (mntss) [inline README.md:147]: This section should go to the `CONTRIBUTING.md` file

File: README.md [hunk 5]
```diff
@@ -116,15 +135,45 @@ circuit-tracer attribute \
...
+You can install circuit-tracer with dev dependencies with `pip install -e ".[dev]"`. This project uses `pytest` for tests, `ruff` for linting and formatting, and `pyright` for type-checking. Before contributing a PR, please run the following to make sure your code passes all checks:
+
+```bash
+python -m ruff check
+python -m ruff format
+python -m pyright
+python -m pytest
+```
+
+If you are using VSCode or Cursor, you can install the ruff plugin, and the add the following to your `.vscode/settings.json` to enable auto-formatting with ruff on save:
```
Comment (mntss) [inline README.md:158]: I'd leave the configuration of the IDE up to the developers

File: .github/workflows/ci.yaml
Comment (mntss) [inline .github/workflows/ci.yaml:1]: Not familiar with GH actions so I might be wrong, but is there a way to make this run as a part of this PR so we can verify that the CI does indeed work?

File: .github/workflows/ci.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+    branches:
+      - main
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
```
Comment (mntss) [inline .github/workflows/ci.yaml:14]: Lets use the 3.11 version declared in `pyproject.toml`

File: tests/transcoder/test_activation_functions.py
Comment (mntss) [inline tests/transcoder/test_activation_functions.py:1]: The new activation tests cover good ground but feel excessive - 10+ test functions for relatively simple math operations. Could be consolidated.

**Review by mntss — COMMENTED** [review]
Unfortunately we hit some bad timing, since we just merged a major release, so we need to update a chunk of this code.
Responding to the comments:

* I think it's better to just fix the typing if we want to enable type checking rather than adding the ignore comments all around the codebase. Happy to assist if you're concerned about breaking some functions
* The JumpReLU indeed accepted a float before, not sure I understand what was broken
* Not sure I follow the comment about dependency management, did you run into any issues?

File: tests/test_attributions_llama.py (hunk 3 above)
Comment (chanind) [reply tests/test_attributions_llama.py:218]: You can run any specific test file with pytest directly using `pytest test/test_attributions_llama.py`. Is there an extra advantage to keeping the main given this functionality is currently provided by pytest? Using a main function this would be run with `python test/test_attributions_llama.py`, which seems like an equivalent command. You can also run individual tests by regex match, e.g. `pytest -k small_llama` which is even less typing and allows more targeting.

File: README.md (hunk 4 above)
Comment (chanind) [reply README.md:147]: will update!

File: .github/workflows/ci.yaml
Comment (chanind) [reply .github/workflows/ci.yaml:1]: yes it should already do that. Since I am not a maintainer and this is my first PR to your repo, usually Github will require you to click a button on the PR page to allow the action to run. This is done as a security precaution since actions can run arbitrary code, but should be harmless in this case since there's no secrets or anything in CI anyway that can be exposed.

File: .github/workflows/ci.yaml (hunk 6 above)
Comment (chanind) [reply .github/workflows/ci.yaml:14]: good call, I'll update this

**Comment by chanind** [comment]
> Unfortunately we hit some bad timing, since we just merged a major release, so we need to update a chunk of this code.

I'll try to rebase and work through what changed, or maybe start from scratch. I probably can't do this until later this week or weekend though

> I think it's better to just fix the typing if we want to enable type checking rather than adding the ignore comments all around the codebase. Happy to assist if you're concerned about breaking some functions

It's not always possible to fix all typing errors, since many errors are results of incorrect or unspecified typing in other libraries (e.g. most of the stuff in ReplacementModel is extending from TLens / torch.nn.Module). We can specify the types of things there as `Any` to effectively disable type checking for those modules, but it feels more targeted to just add `# type: ignore` comments IMO. Happy to change to `Any` if you'd prefer that for this case.

Other common cases are where things can be `None`, but code assumes that the types are not `None`. A common case for this is that `HookedTransformer.tokenizer` is specified as being nullable in TLens, but all code assumes it's not None. We can add `assert model.tokenizer is not None` in the code as another way to get around the error, as that will narrow the typing, but figured the comment approach is less intrusive for your code. Happy to switch this to the assert though!

Some other areas where I wasn't sure the correct typing is around `resources.files` returning a `Traversable`, but the typing for `open()` does not allow `Traversable` as a valid input. It seems like your code works here, so probably the typing on `open()` is incorrect. This could be likely be fixed by changing the code to wrap the Traversable with something like `pathlib.Path(...)`, but figured the least intrusive way to handle this was just to add the comment. 

Other cases where I wasn't sure what the correct type is, for instance `calculate_node_positions()` requires `list[list[Supernode]]`, but `create_graph_visualization()` passes in a `list[Supernode]`. One of these cannot be correct, and I didn't want to make assumptions without running the code somehow to verify which it is.

In general I would slightly push back on the idea that all `# type: ignore` comments are incorrect. Since Python is not a strongly typed language, sometimes libraries do not implement typing in a way that makes it possible for downstream libraries to know exactly what each type is, and situations like this it seems like just ignoring the type checker for that line is more elegant than trying to patch the types.

> The JumpReLU indeed accepted a float before, not sure I understand what was broken

Apologies maybe I should have framed this as a business logic issue more than a type issue. The problem is that JumpReLU threshold is always tensor that's the same shape as `b_enc`, and is thus never a scalar float. It would be like saying that `Transcoder.W_dec` is a float instead of a tensor.

> Not sure I follow the comment about dependency management, did you run into any issues?

No I can install optional dependencies this way, I'm just noting that it is up the to developer to manually manage/switch their venv, python version, manage dev dependencies, etc for the project.

File: tests/test_attributions_gemma.py (hunk 1 above)
Comment (mntss) [reply tests/test_attributions_gemma.py:379]: We shouldn't have to pull anything for these tests - the ones that uses the actual models should be skipped

File: tests/test_attributions_llama.py (hunk 3 above)
Comment (mntss) [reply tests/test_attributions_llama.py:218]: I am aware about that, was referring to debugging the test without setting up the tests in vs code. I don't care much so feel free to remove if this feels important

File: tests/test_attributions_gemma.py (hunk 1 above)
Comment (chanind) [reply tests/test_attributions_gemma.py:379]: ah great, makes things easier

File: tests/test_attributions_llama.py (hunk 3 above)
Comment (chanind) [reply tests/test_attributions_llama.py:218]: You don't have to set up anything in vscode, I usually run pytest from CLI personally so it would be the same as running manually in CLI with a `__main__`

File: .github/workflows/ci.yaml
Comment (chanind) [reply .github/workflows/ci.yaml:1]: strange, maybe I did something wrong, I'll try on my own fork before re-opening this

**Comment by chanind** [comment]
I'm going to start from scratch rather than trying to work through the formatting issues given the new changes. I'll open a new PR soon
